File: leetcode/addTwoNumbers.js
/****************************
题目描述
给出两个 非空 的链表用来表示两个非负的整数。其中，它们各自的位数是按照 逆序 的方式存储的，并且它们的每个节点只能存储 一位 数字。

如果，我们将这两个数相加起来，则会返回一个新的链表来表示它们的和。

您可以假设除了数字 0 之外，这两个数都不会以 0 开头。

示例：

输入：(2 -> 4 -> 3) + (5 -> 6 -> 4)
输出：7 -> 0 -> 8
原因：342 + 465 = 807
题目解析
设立一个表示进位的变量carried，建立一个新链表，把输入的两个链表从头往后同时处理，每两个相加，将结果加上carried后的值作为一个新节点到新链表后面。
 ****************************/

/// 时间复杂度: O(n)
/// 空间复杂度: O(n)

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

TODO: xxxx


function ListNode(val) {
  this.val = val;
  this.next = null;
}

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {};

// 输入：(2 -> 4 -> 3) + (5 -> 6 -> 4)
const l1 = new ListNode(2)
l1.next = new ListNode(4)

File: algorithm/leetcode/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
      head = None
      tail = None
      carry = 0

      while l1 or l2:
        n1 = l1.val if l1 else 0
        n2 = l2.val if l2 else 0
        sum = n1 + n2 + carry

        if not head:
          tail = ListNode(sum % 10)
          head = tail
        else:
          tail.next = ListNode(sum % 10)
          tail = tail.next
        carry = floor(sum / 10)

        if l1:
          l1 = l1.next
        if l2:
          l2 = l2.next
      if carry > 0:
        tail.next = ListNode(carry)
      return head
        